jxl/src/features/patches.rs

```diff
@@ -0,0 +1,290 @@
+const REFERENCE_FRAME_CONTEXT: usize = 1;
+const PATCH_SIZE_CONTEXT: usize = 2;
+const PATCH_REFERENCE_POSITION_CONTEXT: usize = 3;
+const PATCH_POSITION_CONTEXT: usize = 4;
+const PATCH_BLEND_MODE_CONTEXT: usize = 5;
+const PATCH_OFFSET_CONTEXT: usize = 6;
+const PATCH_COUNT_CONTEXT: usize = 7;
+const PATCH_ALPHA_CHANNEL_CONTEXT: usize = 8;
+const PATCH_CLAMP_CONTEXT: usize = 9;
+const NUM_PATCH_DICTIONARY_CONTEXTS: usize = 10;
+
+// Blend modes
+// The new values are the old ones. Useful to skip some channels.
+const PATCH_BLEND_MODE_NONE: u8 = 0;
+// The new values (in the crop) replace the old ones: sample = new
+const PATCH_BLEND_MODE_REPLACE: u8 = 1;
+// The new values (in the crop) get added to the old ones: sample = old + new
+const PATCH_BLEND_MODE_ADD: u8 = 2;
+// The new values (in the crop) get multiplied by the old ones:
+// sample = old * new
+// This blend mode is only supported if BlendColorSpace is kEncoded. The
+// range of the new value matters for multiplication purposes, and its
+// nominal range of 0..1 is computed the same way as this is done for the
+// alpha values in kBlend and kAlphaWeightedAdd.
+const PATCH_BLEND_MODE_MUL: u8 = 3;
+// The new values (in the crop) replace the old ones if alpha>0:
+// For first alpha channel:
+// alpha = old + new * (1 - old)
+// For other channels if !alpha_associated:
+// sample = ((1 - new_alpha) * old * old_alpha + new_alpha * new) / alpha
+// For other channels if alpha_associated:
+// sample = (1 - new_alpha) * old + new
+// The alpha formula applies to the alpha used for the division in the other
+// channels formula, and applies to the alpha channel itself if its
+// blend_channel value matches itself.
+// If using kBlendAbove, new is the patch and old is the original image; if
+// using kBlendBelow, the meaning is inverted.
+const PATCH_BLEND_MODE_BLEND_ABOVE: u8 = 4;
+const PATCH_BLEND_MODE_BLEND_BELOW: u8 = 5;
+// The new values (in the crop) are added to the old ones if alpha>0:
+// For first alpha channel: sample = sample = old + new * (1 - old)
+// For other channels: sample = old + alpha * new
+const PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_ABOVE: u8 = 6;
+const PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_BELOW: u8 = 7;
+const PATCH_BLEND_MODE_NUM_BLEND_MODES: u8 = 8;
+
+#[inline]
+fn uses_alpha(blend_mode: u8) -> bool {
+    blend_mode == PATCH_BLEND_MODE_BLEND_ABOVE
+        || blend_mode == PATCH_BLEND_MODE_BLEND_BELOW
+        || blend_mode == PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_ABOVE
+        || blend_mode == PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_BELOW
+}
+
+#[inline]
+fn uses_clamp(blend_mode: u8) -> bool {
+    uses_alpha(blend_mode) || blend_mode == PATCH_BLEND_MODE_MUL
+}
+
+#[derive(Debug, Clone, Copy)]
+struct PatchBlending {
+    mode: u8,
+    alpha_channel: usize,
+    clamp: bool,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct PatchReferencePosition {
+    // Not using `ref` like in the spec here, because it is a keyword.
+    reference: usize,
```

Comment:
note: you *could* use r#ref, if you wanted (but don't, reference is perfectly fine)

Reply:
ack